### PR TITLE
feat: add hmac-sm3 test cases from GM/T 0042-2015 Appendix D.3

### PR DESCRIPTION
**Pull request by ymj-uos**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/hmactest.c [hunk 1]
```diff
@@ -81,6 +85,62 @@ static struct test_st {
...
+ */
+struct testvec_st {
+    const char key[64];
+    int ksize;
+    const unsigned char plaintext[128];
+    int psize;
+    const char* digest;
+} hmac_sm3_tv[] = {
+    {
+        .key    = "\x01\x02\x03\x04\x05\x06\x07\x08"
```
Comment (paulidale) [inline test/hmactest.c:100]: Our coding style doesn't allow named initialisers.

**Review by paulidale — COMMENTED** [review]
I think this would be better added to evp_test.

File: test/hmactest.c (hunk 1 above)
Comment (ymj-uos) [reply test/hmactest.c:100]: I will fix it

**Comment by ymj-uos** [comment]
> I think this would be better added to evp_test.

which place ?

**Comment by paulidale** [comment]
Look in `test/recipes/30-test_evp_data`.
Either create a new file with these tests or add them to `evpmac_common.txt`

File: test/hmactest.c [hunk 2]
```diff
@@ -25,6 +25,10 @@
 #  include <openssl/md5.h>
 # endif
 
+# ifndef OPENSSL_NO_SM3
```
Comment (paulidale) [inline test/hmactest.c:28]: The changes to this file can be omitted.

File: test/recipes/30-test_evp_data/evpmac_common.txt [hunk 3]
```diff
@@ -239,6 +239,31 @@ Input = "Test that SHAKE128 fails"
 Key = 000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f
 Result = MAC_INIT_ERROR
 
+Title = HMAC-SM3 from GM/T 0042-2015 Appendix D.3
+
+MAC = HMAC
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmac_common.txt:244]: Add ````Availablein = default```` above every MAC = HMAC line..
This makes the tests only run for the default provider. SM3 is not a FIPS algorithm

Comment (slontis) [reply test/recipes/30-test_evp_data/evpmac_common.txt:244]: OR create a separate file evpmac_sm3.txt and add it into 30-test_evp.t at line 108 (which is a list of tests that just run with the default provider)

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
Tests are failing for some reason

**Comment by slontis** [comment]
This PR should not be merged before #18737

**Review by slontis — CHANGES_REQUESTED** [review]
Dependent on #18737 being merged

**Comment by slontis** [comment]
Should this be in 3.0?

**Comment by paulidale** [comment]
Okay with 3.0.  Okay without.

**Comment by ymj-uos** [comment]
> Okay with 3.0. Okay without.


> Okay with 3.0. Okay without.

with what, 
???

**Comment by paulidale** [comment]
Nothing to do here.

**Comment by paulidale** [comment]
Kicking the CIs since #18737 is now merged.

**Comment by paulidale** [comment]
Merged, thanks for the contribution.

**Pull request closed without merging**